### PR TITLE
[Merged by Bors] - ET-3348 fixes possible coi duplication

### PR DESCRIPTION
**Pull request by biernacki**
**References**:
- [ET-3348]

**Summary**:
- added a dedicated table for setting up a lock needed for updating CoIs for every user in a sequential manner

**Before the change**:
- **doc_A** - transaction started
- **doc_A** - CoIs selected for update [CoiId(1)]
- **doc_B** - transaction started
- **doc_A** - now or updated CoI returned from coi calculation CoiId(2) inserted into coi table
- **doc_A** - transaction done
- **doc_B** - CoIs selected for update [CoiId(1)] - **CoI 2 is MISSING**
- **doc_B** - now or updated CoI returned from coi calculation CoiId(3) inserted into coi table
- **doc_B** - transaction done

**After the change**:
- **doc_A** - transaction started
- **doc_A** - CoIs selected for update [CoiId(1)]
- **doc_B** - transaction started
- **doc_A** - now or updated CoI returned from coi calculation CoiId(2) inserted into coi table
- **doc_A** - transaction done
- **doc_B** - CoIs selected for update [CoiId(1), CoiId(2)] - **CoI 2 is included**
- **doc_B** - now or updated CoI returned from coi calculation CoiId(3) inserted into coi table
- **doc_B** - transaction done

[ET-3348]: https://xainag.atlassian.net/browse/ET-3348?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: discovery_engine_core/web-api/src/storage.rs [hunk 1]
```diff
@@ -107,6 +107,15 @@ impl UserState {
...
 
+        sqlx::query("INSERT INTO coi_update_lock (user_id) VALUES ($1) ON CONFLICT DO NOTHING;")
+            .bind(user_id.as_ref())
+            .execute(&mut tx)
+            .await?;
+        sqlx::query("SELECT FROM coi_update_lock WHERE user_id = $1 FOR UPDATE;")
+            .bind(user_id.as_ref())
+            .execute(&mut tx)
+            .await?;
+
```
Comment (biernacki) [inline discovery_engine_core/web-api/src/storage.rs:118]: not sure if we still need the old lock in line 125, below 👇 

```rust
let mut positive_cois: Vec<_> = sqlx::query_as::<_, QueriedCoi>(
  "SELECT coi_id, is_positive, embedding, view_count, view_time_ms, last_view 
    FROM center_of_interest 
    WHERE user_id = $1 AND is_positive 
    FOR UPDATE;", // do we need this line ???
)
```

Comment (rustonaut) [reply discovery_engine_core/web-api/src/storage.rs:118]: shouldn't be necessary

**Review by rustonaut — APPROVED** [review]
lgtm

**Comment by biernacki** [comment]
bors merge

**Pull request closed without merging**